flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/SinkTransformationTranslator.java

```diff
@@ -0,0 +1,293 @@
+				null,
+				"Sink Global Committer:",
+				1,
+				1,
+				sinkTransformation,
+				context);
+	}
+
+	//TODO maybe move to a util class
+	private int getParallelism(
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> sinkTransformation,
+			Context context) {
+		return sinkTransformation.getParallelism() != ExecutionConfig.PARALLELISM_DEFAULT
+				? sinkTransformation.getParallelism()
+				: context.getStreamGraph().getExecutionConfig().getParallelism();
+	}
+
+	/**
+	 * Add a operator to the {@link StreamGraph}.
+	 *
+	 * @param operatorFactory The operator factory
+	 * @param inputId The upstream stream node id of the operator
+	 * @param inTypeInfo The input type information of the operator
+	 * @param outTypInfo The output type information of the operator
+	 * @param prefix The prefix of the name and uid of the operator
+	 * @param parallelism The parallelism of the operator
+	 * @param maxParallelism The max parallelism of the operator
+	 * @param sinkTransformation The sink transformation which the operator belongs to
+	 *
+	 * @return The stream node id of the operator
+	 */
+	private <IN, OUT> int addOperatorToStreamGraph(
+			StreamOperatorFactory<OUT> operatorFactory, int inputId,
+			TypeInformation<IN> inTypeInfo,
+			TypeInformation<OUT> outTypInfo,
+			String prefix,
+			int parallelism,
+			int maxParallelism,
+			SinkTransformation<InputT, CommT, WriterStateT, GlobalCommT> sinkTransformation,
+			Context context) {
+		final StreamGraph streamGraph = context.getStreamGraph();
+		final String slotSharingGroup = context.getSlotSharingGroup();
+		final int transformationId = Transformation.getNewNodeId();
+
+		final ChainingStrategy chainingStrategy = sinkTransformation.getChainingStrategy();
+		if (chainingStrategy != null) {
+			operatorFactory.setChainingStrategy(chainingStrategy);
+		}
+
+		streamGraph.addOperator(
+				transformationId,
+				slotSharingGroup,
+				sinkTransformation.getCoLocationGroupKey(),
+				operatorFactory,
+				inTypeInfo,
+				outTypInfo,
+				String.format("%s %s", prefix, sinkTransformation.getName()));
+
+		streamGraph.setParallelism(transformationId, parallelism);
+		streamGraph.setMaxParallelism(transformationId, maxParallelism);
+
+		StreamGraphUtils.configureBufferTimeout(
+				streamGraph,
+				transformationId,
+				sinkTransformation,
+				context.getDefaultBufferTimeout());
+		if (sinkTransformation.getUid() != null) {
+			streamGraph.setTransformationUID(
+					transformationId,
+					String.format("%s %s", prefix, sinkTransformation.getUid()));
+		}
+		streamGraph.addEdge(inputId, transformationId, 0);
+
+		return transformationId;
+	}
+
+	private CommittableTypeInformation<CommT> extractCommittableTypeInformation(Sink<InputT, CommT, WriterStateT, GlobalCommT> sink) {
+		if (sink.getCommittableSerializer().isPresent()) {
+			final Type committableType = TypeExtractor.getParameterType(
+					Sink.class,
+					sink.getClass(),
+					1);
```

Comment:
Is this whole type extraction logic correct? Is the `1` position correct? @aljoscha could you also have a look on this one?

Reply:
I think it should be correct. But it's easy to verify by just logging at it in the debugger.